### PR TITLE
opt,sql: add support for more geospatial join conditions

### PR DESCRIPTION
**Pull request by rytaft**
Prior to this commit, we only supported perfoming geospatial joins
on a single index-accelerated function. This commit adds support for
a wider variety of join conditions, including ANDs and ORs of multiple
different index-accelerated functions with multiple different input
columns or constant values.

Release note (performance improvement): added support for performing
joins using geospatial inverted indexes in a wider variety of cases.
Complex boolean predicates combining multiple index-accelerated geospatial
functions are now supported as join conditions.

### CODE REVIEW

File: pkg/sql/execinfrapb/processors_sql.proto [hunk 1]
```diff
@@ -513,18 +513,15 @@ message InvertedJoinerSpec {
   // inverted column, and the the remaining columns are the primary key.
   optional uint32 index_idx = 2 [(gogoproto.nullable) = false];
 
-  // Index of the column in the input stream that is to be joined with
-  // the inverted index.
-  optional uint32 lookup_column = 3 [(gogoproto.nullable) = false];
```
Comment (mjibson) [inline pkg/sql/execinfrapb/processors_sql.proto:518]: Mark this id as reserved so it's not reused.

**Review by madelynnblue — COMMENTED** [review]
Wow, this is amazing! So much code simplification and removal! I don't have many comments besides it's unfortunate that the And/Or/Func support has to be added in three places.

File: pkg/sql/execinfrapb/processors_sql.proto (hunk 1 above)
Comment (rytaft) [reply pkg/sql/execinfrapb/processors_sql.proto:518]: Good call - Done.

**Comment by rytaft** [comment]
TFTR! 

> it's unfortunate that the And/Or/Func support has to be added in three places.

Yea, I agree -- but couldn't figure out how to avoid it without making things uglier. Ideas welcome!

**Review by sumeerbhola — CHANGES_REQUESTED** [review]
looks good! just some minor comments.

> Reviewed 23 of 23 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52376#-:-MDzumOZ1l7Y3c660ATp:blj1g6i)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft)

___
*[pkg/sql/opt/invertedidx/geo.go, line 691 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52376#-MDzsJaF0-JQ2m25DS0Z:-MDzsJaF0-JQ2m25DS0_:b791bh9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b740487f34222e274e873c248c7509d9c03f7565/pkg/sql/opt/invertedidx/geo.go#L691)):*
> ```Go
> 			// it for every row.
> 			var spanExpr *invertedexpr.SpanExpression
> 			if d, ok := nonIndexParam.(tree.Datum); ok {
> ```

if there was an expression like `ST_Intersects(left, <constant>) AND ST_Intersects(right, <constant>) AND ST_Intersects(left, right)` would the `ST_Intersects(left, <constant>)` not be part of the join condition?
___
*[pkg/sql/opt/invertedidx/geo.go, line 695 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52376#-MDztT_847ETsjhqlkvS:-MDztT_847ETsjhqlkvT:bade2cu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b740487f34222e274e873c248c7509d9c03f7565/pkg/sql/opt/invertedidx/geo.go#L695)):*
> ```Go
> 			}
> 
> 			return &geoInvertedExpr{
> ```

So is the purpose of `getInvertedExpr` to generate the two fast-paths corresponding to (a) a constant and the right (indexed) column, (b) extracting the relationship from the map so we don't have to reprocess the FuncExpr for each input row?
It would be good to have a comment to clarify the reason for this TypedExpr to TypedExpr transformation.
___
*[pkg/sql/opt/invertedidx/geo_test.go, line 142 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52376#-MDzmVvn014-NnR_2QqZ:-MDzmVvn014-NnR_2Qq_:b-ze5avo) ([raw file](https://github.com/cockroachdb/cockroach/blob/b740487f34222e274e873c248c7509d9c03f7565/pkg/sql/opt/invertedidx/geo_test.go#L142)):*
> ```Go
> 		},
> 		{
> 			// We can constrain either index when the functions are AND-ed.
> ```

how about tests with ANDing, ORing a non-geospatial function?
___
*[pkg/sql/opt/invertedidx/inverted_index_expr.go, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52376#-MDzi-CF35lu74jNk6Vw:-MDzi-CF35lu74jNk6Vx:bprr2et) ([raw file](https://github.com/cockroachdb/cockroach/blob/b740487f34222e274e873c248c7509d9c03f7565/pkg/sql/opt/invertedidx/inverted_index_expr.go#L25)):*
> ```Go
> // NewDatumToInvertedExpr returns a new DatumsToInvertedExpr. Currently there
> // is only one possible implementation returned, geoDatumToInvertedExpr.
> func NewDatumToInvertedExpr(
> ```

nit: s/NewDatum/NewDatums/
___
*[pkg/sql/opt/xform/custom_funcs.go, line 1926 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52376#-MDzo6XFA0W0RgtiBmQk:-MDzo6XFA0W0RgtiBmQl:b3nhwao) ([raw file](https://github.com/cockroachdb/cockroach/blob/b740487f34222e274e873c248c7509d9c03f7565/pkg/sql/opt/xform/custom_funcs.go#L1926)):*
> ```Go
> 		// Geospatial lookup joins are not covering, so we must wrap them in an
> 		// index join.
> 		if scanPrivate.Flags.NoIndexJoin {
> ```

what causes `NoIndexJoin` to sometimes be true -- is it a directive specified in a query?
___
*[pkg/sql/opt/xform/testdata/rules/join, line 2941 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52376#-MDzh0s-1s5elsTqrMNB:-MDzh0s-1s5elsTqrMNC:b-7c7gol) ([raw file](https://github.com/cockroachdb/cockroach/blob/b740487f34222e274e873c248c7509d9c03f7565/pkg/sql/opt/xform/testdata/rules/join#L2941)):*
> ```
>            └── st_covers(n.geom:15, c.geom:10) [outer=(10,15), immutable]
> 
> # Complex join filter.
> ```

Can you add cases for:
- geo-expression AND non-geo-expression
- geo-expression OR non-geo-expression: forcing the use of the geo index should fail.
___
*[pkg/sql/rowexec/inverted_joiner.go, line 157 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52376#-MDzfXp7BHBmGs2vZoUc:-MDzfXp7BHBmGs2vZoUd:bmrbva4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b740487f34222e274e873c248c7509d9c03f7565/pkg/sql/rowexec/inverted_joiner.go#L157)):*
> ```Go
> 	processorID int32,
> 	spec *execinfrapb.InvertedJoinerSpec,
> 	datumToInvertedExpr invertedexpr.DatumsToInvertedExpr,
> ```

nit: parameter name should be changed to `datumsToInvertedExpr`

**Review by rytaft — COMMENTED** [review]
TFTR! Updated.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52376#-:-ME484NIAYTMsJq4mNUo:b-p0fj95)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/sql/opt/invertedidx/geo.go, line 691 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52376#-MDzsJaF0-JQ2m25DS0Z:-ME3IIYT6rjS5-RcN8gN:b-rq6485) ([raw file](https://github.com/cockroachdb/cockroach/blob/b740487f34222e274e873c248c7509d9c03f7565/pkg/sql/opt/invertedidx/geo.go#L691)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

if there was an expression like `ST_Intersects(left, <constant>) AND ST_Intersects(right, <constant>) AND ST_Intersects(left, right)` would the `ST_Intersects(left, <constant>)` not be part of the join condition?
</blockquote></details>

It shouldn't be -- added a test. I had to add another rule to make the optimizer actually consider using an inverted join in this case.
___
*[pkg/sql/opt/invertedidx/geo.go, line 695 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52376#-MDztT_847ETsjhqlkvS:-ME3J76T9Y4U3EDSoIDh:b-4tk5xc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b740487f34222e274e873c248c7509d9c03f7565/pkg/sql/opt/invertedidx/geo.go#L695)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

So is the purpose of `getInvertedExpr` to generate the two fast-paths corresponding to (a) a constant and the right (indexed) column, (b) extracting the relationship from the map so we don't have to reprocess the FuncExpr for each input row?
It would be good to have a comment to clarify the reason for this TypedExpr to TypedExpr transformation.
</blockquote></details>

Exactly - Done.
___
*[pkg/sql/opt/invertedidx/geo_test.go, line 142 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52376#-MDzmVvn014-NnR_2QqZ:-ME3Yb7Q6JHKMggAqFP3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b740487f34222e274e873c248c7509d9c03f7565/pkg/sql/opt/invertedidx/geo_test.go#L142)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

how about tests with ANDing, ORing a non-geospatial function?
</blockquote></details>

Done.
___
*[pkg/sql/opt/invertedidx/inverted_index_expr.go, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52376#-MDzi-CF35lu74jNk6Vw:-ME3YjsABpf_PBCuot7M:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b740487f34222e274e873c248c7509d9c03f7565/pkg/sql/opt/invertedidx/inverted_index_expr.go#L25)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: s/NewDatum/NewDatums/
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/custom_funcs.go, line 1926 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52376#-MDzo6XFA0W0RgtiBmQk:-ME3Ym5a2dTkriG_xNI_:b-b3cb3e) ([raw file](https://github.com/cockroachdb/cockroach/blob/b740487f34222e274e873c248c7509d9c03f7565/pkg/sql/opt/xform/custom_funcs.go#L1926)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

what causes `NoIndexJoin` to sometimes be true -- is it a directive specified in a query?
</blockquote></details>

Yea, it's an index hint, e.g. `table@{NO_INDEX_JOIN}`
___
*[pkg/sql/opt/xform/testdata/rules/join, line 2941 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52376#-MDzh0s-1s5elsTqrMNB:-ME3_Pj6CJRCWgKKRXQS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b740487f34222e274e873c248c7509d9c03f7565/pkg/sql/opt/xform/testdata/rules/join#L2941)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Can you add cases for:
- geo-expression AND non-geo-expression
- geo-expression OR non-geo-expression: forcing the use of the geo index should fail.

</blockquote></details>

Done.
___
*[pkg/sql/rowexec/inverted_joiner.go, line 157 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52376#-MDzfXp7BHBmGs2vZoUc:-ME3_Yv38SyQBp6yJAi-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b740487f34222e274e873c248c7509d9c03f7565/pkg/sql/rowexec/inverted_joiner.go#L157)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: parameter name should be changed to `datumsToInvertedExpr`
</blockquote></details>

Done.

**Review by sumeerbhola — APPROVED** [review]
:lgtm: 

> Reviewed 17 of 17 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52376#-:-ME4wpGF8ZOjVMHDPCV2:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rytaft and @sumeerbhola)

___
*[pkg/sql/opt/xform/testdata/rules/join, line 3127 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52376#-ME4ttnA7VDjs6Pp6-Vj:-ME4ttnA7VDjs6Pp6-Vk:by8u2qv) ([raw file](https://github.com/cockroachdb/cockroach/blob/9294ebdca1acd670924ec0f2c992d255820256f9/pkg/sql/opt/xform/testdata/rules/join#L3127)):*
> ```
>       │    │    ├── index-join nyc_neighborhoods
>       │    │    │    ├── columns: name:14 n.geom:15
>       │    │    │    └── inverted-filter
> ```

I'm trying to understand this plan. It seems that the optimizer decided to use an inverted-filterer with `ST_Covers('LINESTRING ( 0 0, 0 2 )', n.geom)` first. And then use that to do an inverted join with nyc_neighborhoods.
If we did not force `nyc_neighborhoods@nyc_neighborhoods_geo_idx` what happens? The cardinality of the nyc_census_blocks is much higher than nyc_neighborhoods (38794 rows vs 129 rows). I am curious whether the plan will do an inverted-filterer using `ST_Covers('LINESTRING ( 0 0, 0 2 )', n.geom)` and then do an inverted join with nyc_census_blocks.

**Review by rytaft — COMMENTED** [review]
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52376#-:-ME5Glww025bkJ6BhhwD:b-oduykh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/sql/opt/xform/testdata/rules/join, line 3127 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52376#-ME4ttnA7VDjs6Pp6-Vj:-ME5GJtD5Chn_FxUy9ne:b-ihmyky) ([raw file](https://github.com/cockroachdb/cockroach/blob/9294ebdca1acd670924ec0f2c992d255820256f9/pkg/sql/opt/xform/testdata/rules/join#L3127)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I'm trying to understand this plan. It seems that the optimizer decided to use an inverted-filterer with `ST_Covers('LINESTRING ( 0 0, 0 2 )', n.geom)` first. And then use that to do an inverted join with nyc_neighborhoods.
If we did not force `nyc_neighborhoods@nyc_neighborhoods_geo_idx` what happens? The cardinality of the nyc_census_blocks is much higher than nyc_neighborhoods (38794 rows vs 129 rows). I am curious whether the plan will do an inverted-filterer using `ST_Covers('LINESTRING ( 0 0, 0 2 )', n.geom)` and then do an inverted join with nyc_census_blocks.

</blockquote></details>

This plan is already doing an inverted join with `nyc_census_blocks`, which is the same plan that is chosen without the index hint.

**Pull request merged**